web/src/compose_ui.js

```diff
@@ -682,6 +682,89 @@ export function format_text($textarea, type, inserted_content) {
         );
     };
 
+    // Links have to be formatted differently because formatting is not only
+    // at the beginning and end of the text, but also in the middle
+    // Therefore more checks are necessary if selected text is already formatted
+    const format_link = () => {
+        const link_syntax_start = "[";
+        const link_syntax_end = "](url)";
+
+        // Captures:
+        // [<description>](<url>)
+        // with just <url> selected
+        const is_selection_url = () =>
+            range.start >= "[](".length &&
+            text.length - range.end >= ")".length &&
+            text.slice(range.start - 2, range.start) === "](" &&
+            text[range.end] === ")" &&
+            text.lastIndexOf("[", range.start - 3) < text.lastIndexOf("]", range.start - 2);
```

Comment:
This is decently readable. But I think a nicer pattern would be:
```
const is_selection_url = () =>
    has_text_just_after_selection(text, range, ")") &&
    has_text_just_before_selection(text, range, "](") &&
    text.lastIndexOf("[", range.start - 3) < text.lastIndexOf("]", range.start - 2);
```

We can leave that for idea for future refactoring; this is already pretty complex. And I have two other bigs thing I'd want to refactor once we have this working, which I'll post about separately.

Well, one of them is that I don't love how many inline functions this has -- we could just have these be independent functions that accept a few parameters like `(text, range)` -- for more readable code.